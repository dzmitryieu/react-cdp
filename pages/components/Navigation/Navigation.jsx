import React from 'react';
import PropTypes from 'prop-types';
import styles from './Navigation.scss';

const links = {
  'release date': ['release_date', 'releaseDateSort'],
  rating: ['vote_count', 'voteCountSort'],
};

class Navigation extends React.Component {
  state = {
    releaseDateSort: 'DESC',
    voteCountSort: 'DESC',
  }

  onClickHandler = (e) => {
    const { releaseDateSort, voteCountSort } = this.state;
    const { sortFilms } = this.props;
    if (e.target.dataset.sorttype === 'release_date') {
      this.setState({
        releaseDateSort: releaseDateSort === 'ASC' ? 'DESC' : 'ASC',
      });
      sortFilms({
        key: 'release_date',
        direction: releaseDateSort === 'DESC' ? 1 : -1,
      });
    } else {
      this.setState({
        voteCountSort: voteCountSort === 'ASC' ? 'DESC' : 'ASC',
      });
      sortFilms({
        key: 'vote_count',
        direction: voteCountSort === 'DESC' ? 1 : -1,
      });
    }
  }

  render() {
    const { films, searchingWord, typeOfSearch } = this.props;
    return (
      <div className={styles.wrapper}>
        <div className={styles.container}>
          <h3 className={styles.foundTitle}>{searchingWord && `${films.length} movies found by "${searchingWord}" ${typeOfSearch}`}</h3>
          <nav className={styles['nav-links']}>
            <h3 className={styles.sortTitle}>Sort by</h3>
            {Object.keys(links).map(keylink => (
              <button
                key={keylink}
                type="button"
                data-sorttype={links[keylink][0]}
                className={`${styles.link} ${this.state[links[keylink][1]] === 'DESC' ? styles.up : styles.down}`} // eslint-disable-line react/destructuring-assignment
                onClick={this.onClickHandler}
              >
                {keylink}
              </button>
            ))}
          </nav>
        </div>
      </div>
    );
  }
}

Navigation.propTypes = {
  films: PropTypes.arrayOf(PropTypes.object).isRequired,
  sortFilms: PropTypes.func.isRequired,
  searchingWord: PropTypes.string.isRequired,
  typeOfSearch: PropTypes.string.isRequired,
};


export default Navigation;
