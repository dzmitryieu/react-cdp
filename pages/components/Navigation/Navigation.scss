@import "../../../src/utils/constants.scss";

.wrapper {
  width: 100%;
}

.container {
  width: 100%;
  min-height: 80px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  border-bottom: 1px solid $line-color;
  padding: 0 130px;
}

.nav-links {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 400px;
}

.foundTitle {
  font-size: 24px;
  text-align: center;
}

.sortTitle {
  margin: 0 10px;
  font-size: 22px;
  width: 100%;
}

.link {
  position: relative;
  height: 30px;
  color: $light-red;
  font-size: 14px;
  letter-spacing: 1px;
  line-height: 22px;
  font-family: $helvetica;
  margin: 0 10px;
  font-weight: normal;
  cursor: pointer;
  border: none;
  background-color: transparent;
  outline: none;
  z-index: 5;
  width: 100%;

  &::before {
    content: '';
    top: 12px;
    right: 0;
    position: absolute;
    width: 5px;
    height: 5px;
    border-style: solid;
    border-width: 0 2px 2px 0;
    transition: all ease 0.3s;
  }
}

.up.link::before {
  transform: rotate(-135deg);
  top: 12px;
}

.down.link::before {
  transform: rotate(45deg);
  top: 12px;
}

.active {
  font-weight: bold;
  border-bottom: 2px solid $light-red;
}

@media screen and (max-width: $tablet-size-screen) {
  .link {
    font-size: 12px;
    margin: 0 5px;
    line-height: 1;
  }

  .nav-links {
    width: 280px;
  }
}

@media screen and (max-width: $phone-size-screen) {
  .container {
    padding: 0 20px;
  }

  .link {
    font-size: 10px;
  }

  .foundTitle {
    font-size: 18px;
  }

  .sortTitle {
    text-align-last: left;
    font-size: 14px;
  }

  .nav-links {
    width: 200px;
  }
}
