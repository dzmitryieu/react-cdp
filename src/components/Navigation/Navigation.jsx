import React from 'react';
import styles from './Navigation.scss';

const links = ['release date', 'rating'];

const Navigation = () => (
  <div className={styles.wrapper}>
    <div className={styles.container}>
      <h3 className={styles.foundTitle}>9 movies found</h3>
      <nav className={styles['nav-links']}>
        <h3 className={styles.sortTitle}>Sort by</h3>
        {links.map(link => (
          <button
            key={link}
            type="button"
            className={`${styles.link}`}
          >
            {link}
          </button>
        ))}
      </nav>
    </div>
  </div>
);


export default Navigation;
