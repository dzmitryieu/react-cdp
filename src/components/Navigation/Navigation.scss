@import "../../utils/constants.scss";

.wrapper {
  width: 100%;
}

.container {
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
  border-bottom: 1px solid $line-color;
  padding: 0 50px;
}

.nav-links {
  display: flex;
  justify-content: center;
  align-items: center;
}

.sortTitle {
  margin: 0 10px;
}

.link {
  height: 30px;
  color: $light-red;
  font-size: 14px;
  letter-spacing: 1px;
  line-height: 22px;
  font-family: $helvetica;
  margin: 0 10px;
  font-weight: normal;
  cursor: pointer;
  border: none;
  background-color: transparent;
  outline: none;
}

.active {
  font-weight: bold;
  border-bottom: 2px solid $light-red;
}

@media screen and (max-width: $tablet-size-screen) {
  .link {
    font-size: 12px;
  }
}

@media screen and (max-width: $phone-size-screen) {
  .view-icons {
    display: none;
  }

  .link {
    font-size: 10px;
  }
}
