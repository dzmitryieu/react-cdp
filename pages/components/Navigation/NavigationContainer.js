import { connect } from 'react-redux';
import Navigation from './Navigation';
import { sortFilms } from '../../../src/modules/films/filmsActions';
import { getFilms } from '../../../src/modules/films/filmsSelectors';
import { getTypeOfSearch, getSearchingWord } from '../../../src/modules/searchFilms/searchFilmsSelectors';

const mapStateToProps = state => ({
  films: getFilms(state),
  typeOfSearch: getTypeOfSearch(state),
  searchingWord: getSearchingWord(state),
});

const mapDispatchToProps = {
  sortFilms,
};

export default connect(mapStateToProps, mapDispatchToProps)(Navigation);
