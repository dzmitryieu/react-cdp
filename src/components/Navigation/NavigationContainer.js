import { connect } from 'react-redux';
import Navigation from './Navigation';
import { sortFilms } from '../../modules/films/filmsActions';
import { getFilms } from '../../modules/films/filmsSelectors';
import { getTypeOfSearch, getSearchingWord } from '../../modules/searchFilms/searchFilmsSelectors';

const mapStateToProps = state => ({
  films: getFilms(state),
  typeOfSearch: getTypeOfSearch(state),
  searchingWord: getSearchingWord(state),
});

const mapDispatchToProps = {
  sortFilms,
};

export default connect(mapStateToProps, mapDispatchToProps)(Navigation);
