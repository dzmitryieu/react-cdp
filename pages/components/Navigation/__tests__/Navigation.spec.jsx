import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Testrenderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from 'react-redux';
import Navigation from '../Navigation';
import NavigationContainer from '../index';

const renderer = new ShallowRenderer();
Enzyme.configure({ adapter: new Adapter() });

const films = [
  {
    id: '1345647456',
    title: 'Captain Marvel',
    poster_path: '../../assets/images/poster.jpg',
    genres: ['Drama', 'Comedy'],
    release_date: '1984-12-03',
  },
  {
    id: '1345678696',
    title: 'Captain Marvel',
    poster_path: '../../assets/images/poster.jpg',
    genres: ['Drama', 'Comedy'],
    release_date: '1984-12-03',
  },
  {
    id: '13324256',
    title: 'Captain Marvel',
    poster_path: '../../assets/images/poster.jpg',
    genres: ['Drama', 'Comedy'],
    release_date: '1984-12-03',
  },
  {
    id: '34613456',
    title: 'Captain Marvel',
    poster_path: '../../assets/images/poster.jpg',
    genres: ['Drama', 'Comedy'],
    release_date: '1984-12-03',
  },
  {
    id: '1378970456',
    title: 'Captain Marvel',
    poster_path: '../../assets/images/poster.jpg',
    genres: ['Drama', 'Comedy'],
    release_date: '1984-12-03',
  },
];

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore({
  films:
    {
      allFilms: films,
    },
  search:
    {
      searchBy: 'title',
      searchedWord: 'terminator',
    },
});
const sortFilms = jest.fn();
const typeOfSearch = 'title';
const searchingWord = 'Terminator';

describe('Navigation Snapshot', () => {
  test('renders', () => {
    const component = renderer.render(
      <Navigation
        films={films}
        sortFilms={sortFilms}
        searchingWord={searchingWord}
        typeOfSearch={typeOfSearch}
      />,
    );
    expect(component).toMatchSnapshot();
  });
  test('with store render', () => {
    const component = Testrenderer.create(
      <Provider store={store}>
        <NavigationContainer />
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });
});

describe('Navigation links should work', () => {
  it('click to the release_date link should sort array of films', () => {
    const wrapper = shallow(
      <Navigation
        films={films}
        sortFilms={sortFilms}
        searchingWord={searchingWord}
        typeOfSearch={typeOfSearch}
      />,
    );
    const e = {
      target: {
        dataset: {
          sorttype: 'release_date',
        },
      },
    };
    wrapper.find('[data-sorttype="release_date"]').prop('onClick')(e);
    expect(wrapper.state().releaseDateSort).toEqual('ASC');
  });
  it('double click to the release_date link should sort array of films again', () => {
    const wrapper = shallow(
      <Navigation
        films={films}
        sortFilms={sortFilms}
        searchingWord={searchingWord}
        typeOfSearch={typeOfSearch}
      />,
    );
    const e = {
      target: {
        dataset: {
          sorttype: 'release_date',
        },
      },
    };
    wrapper.find('[data-sorttype="release_date"]').prop('onClick')(e);
    wrapper.find('[data-sorttype="release_date"]').prop('onClick')(e);
    expect(wrapper.state().releaseDateSort).toEqual('DESC');
  });
  it('click to the rating link should sort array of films', () => {
    const wrapper = shallow(
      <Navigation
        films={films}
        sortFilms={sortFilms}
        searchingWord={searchingWord}
        typeOfSearch={typeOfSearch}
      />,
    );
    const e = {
      target: {
        dataset: {
          sorttype: 'rating',
        },
      },
    };
    wrapper.find('[data-sorttype="release_date"]').prop('onClick')(e);
    expect(wrapper.state().voteCountSort).toEqual('ASC');
  });
  it('double click to the rating link should sort array of films again', () => {
    const wrapper = shallow(
      <Navigation
        films={films}
        sortFilms={sortFilms}
        searchingWord={searchingWord}
        typeOfSearch={typeOfSearch}
      />,
    );
    const e = {
      target: {
        dataset: {
          sorttype: 'rating',
        },
      },
    };
    wrapper.find('[data-sorttype="release_date"]').prop('onClick')(e);
    wrapper.find('[data-sorttype="release_date"]').prop('onClick')(e);
    expect(wrapper.state().voteCountSort).toEqual('DESC');
  });
});
