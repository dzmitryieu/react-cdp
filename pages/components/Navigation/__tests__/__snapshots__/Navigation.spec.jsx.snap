// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Snapshot renders 1`] = `
<div
  className="wrapper"
>
  <div
    className="container"
  >
    <h3
      className="foundTitle"
    >
      5 movies found by "Terminator" title
    </h3>
    <nav
      className="nav-links"
    >
      <h3
        className="sortTitle"
      >
        Sort by
      </h3>
      <button
        className="link up"
        data-sorttype="release_date"
        onClick={[Function]}
        type="button"
      >
        release date
      </button>
      <button
        className="link up"
        data-sorttype="vote_count"
        onClick={[Function]}
        type="button"
      >
        rating
      </button>
    </nav>
  </div>
</div>
`;

exports[`Navigation Snapshot with store render 1`] = `
<div
  className="wrapper"
>
  <div
    className="container"
  >
    <h3
      className="foundTitle"
    >
      5 movies found by "terminator" title
    </h3>
    <nav
      className="nav-links"
    >
      <h3
        className="sortTitle"
      >
        Sort by
      </h3>
      <button
        className="link up"
        data-sorttype="release_date"
        onClick={[Function]}
        type="button"
      >
        release date
      </button>
      <button
        className="link up"
        data-sorttype="vote_count"
        onClick={[Function]}
        type="button"
      >
        rating
      </button>
    </nav>
  </div>
</div>
`;
